# TMDB 5000 Movies Data Analytics 

## Data Loading:

In [149]:
import numpy as np
import pandas as pd

In [128]:
df = pd.read_csv('datasets/tmdb-5000-movie-dataset/tmdb_5000_movies.csv')

## Questions:

### 1) Which genre of movies is most popular?

In [129]:
def distinct_counts(dataframe, col_name):
    return dataframe[col_name].value_counts()

In [130]:
def convert_json_to_df(dataframe, col_name):
    jsdf = pd.DataFrame()
    for item in dataframe[col_name]:
        js = pd.read_json(item)
        jsdf = jsdf.append(js)
    return jsdf

In [131]:
distinct_counts(df,"genres").head(10)

[{"id": 18, "name": "Drama"}]                                                                    370
[{"id": 35, "name": "Comedy"}]                                                                   282
[{"id": 18, "name": "Drama"}, {"id": 10749, "name": "Romance"}]                                  164
[{"id": 35, "name": "Comedy"}, {"id": 10749, "name": "Romance"}]                                 144
[{"id": 35, "name": "Comedy"}, {"id": 18, "name": "Drama"}]                                      142
[{"id": 35, "name": "Comedy"}, {"id": 18, "name": "Drama"}, {"id": 10749, "name": "Romance"}]    109
[{"id": 27, "name": "Horror"}, {"id": 53, "name": "Thriller"}]                                    88
[{"id": 99, "name": "Documentary"}]                                                               68
[{"id": 27, "name": "Horror"}]                                                                    64
[{"id": 18, "name": "Drama"}, {"id": 53, "name": "Thriller"}]                              

### Drama-only movies top the list.

In [132]:
genres = convert_json_to_df(df,"genres")
distinct_counts(genres,"name").head(10)

Drama              2297
Comedy             1722
Thriller           1274
Action             1154
Romance             894
Adventure           790
Crime               696
Science Fiction     535
Horror              519
Family              513
Name: name, dtype: int64

### Drama is the most viewed genre and TV Movie types are least popular.

### 2) In which movie language were majority of the films released?

In [133]:
distinct_counts(df,"original_language").head(10)

en    4505
fr      70
es      32
de      27
zh      27
hi      19
ja      16
it      14
cn      12
ko      11
Name: original_language, dtype: int64

In [134]:
lang_df = df[['original_language', 'spoken_languages']].copy()
unique_lang_df = pd.DataFrame(lang_df.groupby(["original_language", "spoken_languages"]).size().reset_index(name="frequency"))
unique_lang_df = unique_lang_df.sort_values(by=['frequency'], ascending=False)
unique_lang_df.head(10)

original_language                                   spoken_languages  \
367                en           [{"iso_639_1": "en", "name": "English"}]   
180                en  [{"iso_639_1": "en", "name": "English"}, {"iso...   
244                en  [{"iso_639_1": "en", "name": "English"}, {"iso...   
33                 en                                                 []   
374                en  [{"iso_639_1": "es", "name": "Espa\u00f1ol"}, ...   
155                en  [{"iso_639_1": "en", "name": "English"}, {"iso...   
288                en  [{"iso_639_1": "en", "name": "English"}, {"iso...   
339                en  [{"iso_639_1": "en", "name": "English"}, {"iso...   
513                fr     [{"iso_639_1": "fr", "name": "Fran\u00e7ais"}]   
301                en  [{"iso_639_1": "en", "name": "English"}, {"iso...   

     frequency  
367       3153  
180        123  
244        108  
33          85  
374         52  
155         52  
288         50  
339         49  
513         38  
301         22

### Majority of the films were released in English.

### 3) Is there a connection between release date and film's success?

In [135]:
day_list = []
for item in df['release_date']:
    date_object = pd.to_datetime(item)
    day = date_object.day_name()
    day_list.append(day)

In [136]:
weekday = pd.DataFrame(day_list)
weekday[0].value_counts()

Friday       2053
Thursday      916
Wednesday     795
Tuesday       351
Saturday      237
Monday        233
Sunday        217
Name: 0, dtype: int64

### Yes, most of the top movies were released on Fridays.

### 4) What were the most profitable movies?

In [137]:
profit = df[['original_title','budget','revenue']].copy()
profit.head(5)

original_title     budget     revenue
0                                    Avatar  237000000  2787965087
1  Pirates of the Caribbean: At World's End  300000000   961000000
2                                   Spectre  245000000   880674609
3                     The Dark Knight Rises  250000000  1084939099
4                               John Carter  260000000   284139100

In [138]:
profit['profit'] = profit['revenue'] - profit['budget']
profit = profit.sort_values(by=['profit'], ascending=False)
profit.head(10)

original_title     budget     revenue  \
0                                           Avatar  237000000  2787965087   
25                                         Titanic  200000000  1845034188   
28                                  Jurassic World  150000000  1513528810   
44                                       Furious 7  190000000  1506249360   
16                                    The Avengers  220000000  1519557910   
7                          Avengers: Age of Ultron  280000000  1405403694   
124                                         Frozen  150000000  1274219009   
546                                        Minions   74000000  1156730962   
329  The Lord of the Rings: The Return of the King   94000000  1118888979   
31                                      Iron Man 3  200000000  1215439994   

         profit  
0    2550965087  
25   1645034188  
28   1363528810  
44   1316249360  
16   1299557910  
7    1125403694  
124  1124219009  
546  1082730962  
329  1024888979  
31   1015439994

### Avatar made the highest profit of 2550965087.

### 5) Did runtime have an impact on vote average?

In [139]:
runtime = df[['original_title','runtime','vote_average']].copy()
runtime = runtime[runtime.vote_average != 0.0]
runtime = runtime.sort_values(by='vote_average', ascending = False)
runtime.head(10)

original_title  runtime  vote_average
4045     Dancer, Texas Pop. 81     97.0          10.0
4247     Me You and Five Bucks     90.0          10.0
3519          Stiff Upper Lips     99.0          10.0
4662            Little Big Top      0.0          10.0
3992                 Sardaarji      0.0           9.5
2386            One Man's Hero    121.0           9.3
2970        There Goes My Baby     99.0           8.5
1881  The Shawshank Redemption    142.0           8.5
2796     The Prisoner of Zenda    101.0           8.4
3337             The Godfather    175.0           8.4

### There is no connection between runtime and vote average

### 6) Which production companies produced majority of the films? Which company produced successful films?

In [140]:
prod_companies = df[['production_companies']].copy()
prod_companies_df = convert_json_to_df(prod_companies,'production_companies')
distinct_counts(prod_companies_df,"name").head(10)

Warner Bros.                              319
Universal Pictures                        311
Paramount Pictures                        285
Twentieth Century Fox Film Corporation    222
Columbia Pictures                         201
New Line Cinema                           165
Metro-Goldwyn-Mayer (MGM)                 122
Touchstone Pictures                       118
Walt Disney Pictures                      114
Relativity Media                          102
Name: name, dtype: int64

### Warner Bros. produced majority of the films.

In [141]:
success_companies = profit.copy()
success_companies['production_companies']=df['production_companies']
success_companies = success_companies[(success_companies.profit > 0) & (success_companies.budget > 0) & (success_companies.revenue > 0)]
success_companies.sort_values(by='profit', ascending=False)
success_companies_df = convert_json_to_df(success_companies,'production_companies')
distinct_counts(success_companies_df,"name").head(10)

Universal Pictures                        230
Warner Bros.                              219
Paramount Pictures                        202
Twentieth Century Fox Film Corporation    175
Columbia Pictures                         142
New Line Cinema                           121
Walt Disney Pictures                       87
Relativity Media                           84
Metro-Goldwyn-Mayer (MGM)                  67
Touchstone Pictures                        64
Name: name, dtype: int64

### Universal Pictures produced more number of profitable films

### 7) Does popularity depend on vote count?

In [142]:
popularity = df[['popularity','vote_count']].copy()
popularity = popularity[popularity.vote_count>0]
popularity = popularity.sort_values(by='vote_count', ascending = False)
popularity.head(10)

popularity  vote_count
96   167.583710       13752
65   187.322927       12002
0    150.437577       11800
16   144.448633       11776
788  514.569956       10995
95   724.247784       10867
287   82.121691       10099
94   481.098624        9742
426   68.550698        9455
127  434.278564        9427

### Popularity does not depend on vote count.

### 8) What are the most popular tags? Are those related with genres?

In [143]:
tags = df[['keywords']].copy()
tags_df = convert_json_to_df(tags,'keywords')
distinct_counts(tags_df,"name").head(10)

woman director          324
independent film        318
duringcreditsstinger    307
based on novel          197
murder                  189
aftercreditsstinger     170
violence                150
dystopia                139
sport                   126
revenge                 118
Name: name, dtype: int64

### Movies directed by woman are most popular.  

In [144]:
keywordsdf = df[['genres','keywords']].copy()
keywordsdf = keywordsdf.replace('[]', np.nan)
keywordsdf = keywordsdf.dropna()
unique_keywordsdf = pd.DataFrame(keywordsdf.groupby(['genres','keywords']).size().reset_index(name="frequency"))
unique_keywordsdf = unique_keywordsdf.sort_values(by=['frequency'], ascending=False)
unique_keywordsdf.head(10)

genres  \
1512                      [{"id": 18, "name": "Drama"}]   
827   [{"id": 18, "name": "Drama"}, {"id": 10749, "n...   
1567                      [{"id": 18, "name": "Drama"}]   
3481                     [{"id": 35, "name": "Comedy"}]   
3078  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4281                [{"id": 99, "name": "Documentary"}]   
3537                     [{"id": 35, "name": "Comedy"}]   
2853  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
2872  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
1511                      [{"id": 18, "name": "Drama"}]   

                                               keywords  frequency  
1512        [{"id": 10183, "name": "independent film"}]         11  
827         [{"id": 10183, "name": "independent film"}]          8  
1567         [{"id": 187056, "name": "woman director"}]          8  
3481        [{"id": 10183, "name": "independent film"}]          7  
3078        [{"id": 10183, "name": "independent film"}]          7  
4281         [{"id": 187056, "name": "woman director"}]          5  
3537   [{"id": 179431, "name": "duringcreditsstinger"}]          5  
2853        [{"id": 10183, "name": "independent film"}]          4  
2872         [{"id": 187056, "name": "woman director"}]          4  
1511  [{"id": 10183, "name": "independent film"}, {"...          4

### Drama and Independent film combination is widely viewed.

### 9) Which countries earn most profit?

In [145]:
production_countries = profit.copy()
production_countries['production_countries'] = df['production_countries']
production_countries = production_countries[production_countries.profit>0]
production_countries.sort_values(by='profit', ascending = False)
production_countries_json = convert_json_to_df(production_countries,"production_countries")
distinct_counts(production_countries_json,"name").head(10)

United States of America    2339
United Kingdom               343
Germany                      165
France                       140
Canada                       120
Australia                     61
Spain                         35
China                         34
Italy                         30
Japan                         30
Name: name, dtype: int64

### United States of America earns the most profit.

### 10) On an average, how many other language films were released per year?

In [146]:
other_lang = df[['spoken_languages','release_date']].copy()
other_lang['year'] = pd.DatetimeIndex(other_lang['release_date']).year
other_lang = other_lang.fillna(0)
other_lang.year = other_lang.year.astype(int)
other_lang = other_lang.replace('[]', np.nan)
other_lang = other_lang.dropna()
other_lang = other_lang[(other_lang.spoken_languages != '[{"iso_639_1": "en", "name": "English"}]')]
unique_other_lang = pd.DataFrame(other_lang.groupby(["spoken_languages", "year"]).size().reset_index(name="frequency"))
unique_other_lang = unique_other_lang.sort_values(by='frequency', ascending = False)
unique_other_lang.head(10)

spoken_languages  year  frequency
467  [{"iso_639_1": "en", "name": "English"}, {"iso...  2000         11
318  [{"iso_639_1": "en", "name": "English"}, {"iso...  2008          9
316  [{"iso_639_1": "en", "name": "English"}, {"iso...  2006          8
315  [{"iso_639_1": "en", "name": "English"}, {"iso...  2005          8
473  [{"iso_639_1": "en", "name": "English"}, {"iso...  2006          8
312  [{"iso_639_1": "en", "name": "English"}, {"iso...  2002          8
476  [{"iso_639_1": "en", "name": "English"}, {"iso...  2009          8
317  [{"iso_639_1": "en", "name": "English"}, {"iso...  2007          7
319  [{"iso_639_1": "en", "name": "English"}, {"iso...  2009          7
469  [{"iso_639_1": "en", "name": "English"}, {"iso...  2002          7

In [147]:
jsdf = pd.DataFrame()
for row in other_lang.itertuples(index=True):
    js = pd.read_json(getattr(row, "spoken_languages"))
    js['year'] = getattr(row, "year")
    jsdf = jsdf.append(js)
jsdf.head(10)

iso_639_1      name  year
0        en   English  2009
1        es   Español  2009
0        fr  Français  2015
1        en   English  2015
2        es   Español  2015
3        it  Italiano  2015
4        de   Deutsch  2015
0        en   English  2007
1        fr  Français  2007
0        en   English  2006

In [148]:
jsdf = jsdf[jsdf.name != 'English']
unique_jsdf = pd.DataFrame(jsdf.groupby(["name", "year"]).size().reset_index(name="frequency"))
unique_jsdf = unique_jsdf.sort_values(by='frequency', ascending = False)
unique_jsdf.head(10)

name  year  frequency
209  Français  2009         33
203  Français  2003         25
205  Français  2005         24
202  Français  2002         24
147   Español  2002         22
150   Español  2005         22
206  Français  2006         22
152   Español  2007         21
204  Français  2004         21
208  Français  2008         21

### French and Espanol films are more in number in the other language category.